File: content/problems/0295-find-median-from-data-stream/solutions/two-heaps/Solution.kt
class MedianFinder {
    private val lower = PriorityQueue<Int>(compareByDescending { it })
    private val upper = PriorityQueue<Int>()
    fun addNum(num: Int) { lower.add(num); upper.add(lower.remove()); if (upper.size > lower.size) lower.add(upper.remove()) }
    fun findMedian(): Double = if (lower.size > upper.size) lower.element().toDouble() else (lower.element().toLong() + upper.element()) / 2.0
}

File: content/problems/0295-find-median-from-data-stream/solutions/two-heaps/solution.cpp
class MedianFinder {
    priority_queue<int> lower;
    priority_queue<int, vector<int>, greater<int>> upper;
public:
    MedianFinder() {}
    void addNum(int num) { lower.push(num); upper.push(lower.top()); lower.pop(); if (upper.size() > lower.size()) { lower.push(upper.top()); upper.pop(); } }
    double findMedian() { return lower.size() > upper.size() ? lower.top() : ((long long)lower.top() + upper.top()) / 2.0; }
};
